src/mbgl/style/expression/type_check.cpp

```diff
@@ -0,0 +1,121 @@
+#include <mbgl/style/expression/type_check.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using namespace mbgl::style::expression::type;
+
+TypecheckResult typecheck(const Type& expected, const std::unique_ptr<Expression>& e) {
```

Comment:
You can translate from compile time C++ types to runtime expression types via template specialization. You're already doing this for values with `valueTypeToExpressionType`, and it's also possible to write `functionTypeToSignature`:

```C++
template <class Result, class... Inputs>
type::Signature functionTypeToSignature(Result (*fn) (Inputs...) = nullptr) {
    return Signature(valueTypeToExpressionType<Result>(),
                     valueTypeToExpressionType<Inputs>()...);
}

// Can be used with deduced template parameters:
//   functionTypeToSignature([] (bool, float) -> float { return 0.0; });
// or explicitly specified:
//   functionTypeToSignature<float, bool, float>();
```

(Again, just a sketch. Haven't tried to compile this.)